### ISSUE TITLE
YAML syntax in Fenced code block?

### ISSUE THREAD

**Issue by ickc** [sergiocorreia/panflute#18]
As mentioned in #12 about an enhanced version of csv2table in [pandoc-table-csv-test/panflute-csv2table.ipynb at master · ickc/pandoc-table-csv-test](https://github.com/ickc/pandoc-table-csv-test/blob/master/ipynb/panflute-csv2table.ipynb), I have some questions on the YAML syntax required in the fenced code block.

First, it doesn't allow an opening `---` row, e.g.

```markdown
~~~csv
---
table-width: 1.2
...
**_Fruit_**,~~Price~~,_Number_,`Advantages`
*Bananas~1~*,$1.34,12~units~,"Benefits of eating bananas 
(**Note the appropriately
rendered block markdown**):    

- _built-in wrapper_        
- ~~**bright color**~~

"
*Oranges~2~*,$2.10,5^10^~units~,"Benefits of eating oranges:

- **cures** scurvy
- `tasty`"
~~~
```

will treat the `table-width: 1.2` after the first `---` as the actual table (i.e. not YAML), not the content below.

Second, if there's no YAML like this:

```
~~~csv
**_Fruit_**,~~Price~~,_Number_,`Advantages`
*Bananas~1~*,$1.34,12~units~,"Benefits of eating bananas 
(**Note the appropriately
rendered block markdown**):    

- _built-in wrapper_        
- ~~**bright color**~~

"
*Oranges~2~*,$2.10,5^10^~units~,"Benefits of eating oranges:

- **cures** scurvy
- `tasty`"
~~~
```

It will have an error:

```bash
yaml.scanner.ScannerError: while scanning an alias
  in "<unicode string>", line 1, column 1:
    **_Fruit_**,~~Price~~,_Number_,` ... 
```

i.e. empty YAML still requires a `---`/`...` row to start with.

**Comment by sergiocorreia** [comment]
From [these lines](https://github.com/sergiocorreia/panflute/blob/master/panflute/tools.py#L224), it seems the YAML filters are more focused on the YAML part than on the a data part:

- It assumes that all content is YAML until `---` / `...`, so it doesn't allow opening rows (because they are not needed)
- If no `---` or `...` are found, it assumes that all the content is YAML (this explains your 2nd error)

The reason for this is that I was often working with external CSV files, so my markdown was usually:

```
Lorem ipsum

~~~ csv
source: somefile.csv
title: Some Title
~~~
```

Now, I agree that in your case and without a YAML part it looks uglier:

~~~ csv
---
header1 | header2
cell1 | cell2
~~~

If you think there's a way to improve the filter without breaking compatibility, I would be more than glad to add it.

**Comment by ickc** [comment]
### Suggestion: `panflute.fenced_yaml_filter`?

I don't think there's a natural way to change it without breaking backward compatibility. Do you think it's ok to provide an alternative version of the yaml filter, say, `panflute.fenced_yaml_filter`, that *requires* both an opening row of `---` and a closing row of either `---`/`...`, that can be at arbitrary location (much like what pandoc is currently doing). i.e.

	~~~csv
	---
	source: somefile.csv
	title: Some Title
	...
	index,data
	1,2
	~~~

or

	~~~csv
	index,data
	1,2
	---
	source: somefile.csv
	title: Some Title
	...
	~~~

or

	~~~csv
	index,data
	1,2
	~~~

Are all perfectly fine. For this,

	~~~csv
	index,data
	---
	source: somefile.csv
	title: Some Title
	...
	1,2
	~~~

I'm not sure. Probably we can allow this to happen (basically extracting all YAMLs and the remaining becomes the data.

In the case of

	~~~csv
	index,data
	---
	source: somefile.csv
	...
	1,2
	---
	title: Some Title
	...
	~~~

Even more troublesome, but in the spirit of the current pandoc behavior (that YAML can appears anywhere), this should also be allowed.

### Other questions

1. Currently the YAML filter returns a string, even if the format is, say, `0.25,0.25,0.25,0.25`. It seems that it is not really parsing the YAML (otherwise it should be a list)? And even better, it might try to convert to float whenever possible. Currently, I use `.split` and `float()` to do the conversion. (and for alignment I need to use `.strip` to remove white spaces.) I think it would be better if the data are properly parsed.

2. In your example above, you used `source: somefile.csv`. Did you have another version of panflute CSV filter? I don't seem to find it from your examples in this repo or panflute-filters. By the way, I intended to write my csv filters to be backward compatible to yours (and may even be compatible with pandoc-placetable and pandoc-csv2table, depending on the needs).

**Comment by ickc** [comment]
Forget about the part about trying to convert it to float automatically. That probably is a waste of time. It should be done when needed. But I still think the proper parsing it into a list should be the default.

**Comment by ickc** [comment]
My bad, forget about the proper parsing YAML thing. My YAML syntax is wrong... Sorry!

**Comment by sergiocorreia** [comment]
I think I can add an optional argument that triggers it and that you can call from your filter. Something like `strict=True`, to have standard YAML delimiters

About your other question: I also work with a program called Stata, which outputs tables as .tex chunks, which I then format via a filter. So the source case is not exactly a CSV filter.

**Comment by ickc** [comment]
That will do, sounds good to me!

And do you want my csv filter to be backward compatible with your csv example? I want to have default header = True and calling Title caption instead (placetable and csv2table also did that). Mine have a markdown metadata and I want it default to True as well (psrsing markdown content). etc.

**Comment by sergiocorreia** [comment]
I just fixed this with e6736231381 but used the wrong issue #, closing now. I put some notes on the commit msg btw, but basically you now have a strict mode that you can invoke easily when writing the filter.
